src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+   * @param {!Element} element
+   */
+  declareLayer(element) {
+    this.declareLayer_(element);
+  }
+
+  /**
+   * Eagerly creates a Layer for the element.
+   * @param {!Element} element
+   * @return {!LayoutElement}
+   */
+  declareLayer_(element) {
+    const layout = this.add(element);
+    layout.declareLayer();
+    return layout;
+  }
+
+  onResize_() {
+    const layouts = this.layouts_;
+    for (let i = 0; i < layouts.length; i++) {
+      const layout = layouts[i];
+      layout.undeclareLayer();
+      layout.forgetParentLayer();
+    }
+  }
+
+  scrolled_(element) {
+    let layout = LayoutElement.forOptional(element);
+    if (layout && layout.isLayer()) {
+      layout.requestScrollRemeasure();
+    } else {
+      layout = this.declareLayer_(element);
+    }
+
+    if (this.onScroll_) {
+      this.onScroll_(/* layer.getElements() */);
+    }
+  }
+
+  onScroll(handler) {
+    this.onScroll_ = handler;
+  }
+}
+
+export class LayoutElement {
+  constructor(element) {
+    element[LAYOUT_PROP] = this;
+    /**
+     * @private @const {!Element}
+     */
+    this.element_ = element;
+
+    /** @type {?LayoutElement|undefined} */
+    this.parentLayer_ = undefined;
+
+    this.needsRemeasure_ = true;
+    this.size_ = SizeWh(0, 0);
+    this.position_ = PositionLt(0, 0);
+
+    this.isLayer_ = false;
+    this.isRootLayer_ = false;
+    this.needsScrollRemeasure_ = false;
+    this.scrollLeft_ = 0;
+    this.scrollTop_ = 0;
+
+    this.children_ = [];
+  }
+
+  /**
+   * Gets the element's LayoutElement instance.
+   * @param {!Element}
+   * @return {!LayoutElement}
+   */
+  static for(element) {
+    return dev().assert(LayoutElement.forOptional(element));
+  }
+
+  /**
+   * Gets the element's LayoutElement instance, if the element has one.
+   * @param {!Element}
+   * @return {?LayoutElement}
+   */
+  static forOptional(element) {
+    return element[LAYOUT_PROP];
+  }
+
+  /**
+   * Finds the element's parent layer
+   * If the element is itself a layer, it returns the layer's parent layer.
+   * @param {!Element}
+   * @param {opt_force=} Whether to force a re-lookup
+   * @return {?LayoutElement}
+   */
+  static getParentLayer(element, opt_force) {
+    if (!opt_force) {
+      const layout = LayoutElement.forOptional(element);
+      if (layout) {
+        return layout.getParentLayer();
+      }
+    }
+
+    const win = element.ownerDocument.defaultView;
+    if (computedStyle(win, element).position == 'fixed') {
+      LayoutLayers.declareLayer(element);
+      return null;
+    }
+
+    let last = element;
+    let op = element.offsetParent;
+    for (let el = last.parentNode; el; el = el.parentNode) {
+      const layout = LayoutElement.forOptional(el);
+      if (layout && layout.isLayer()) {
+        return layout;
+      }
+
+      // now check to see if offsetParent is a fixed layer
+      if (el === op) {
+        if (computedStyle(win, op).position == 'fixed') {
```

Comment:
Re: style computation cost, can we store the `CSSStyleDeclaration` instead?

> The returned style is a live CSSStyleDeclaration object, which updates itself automatically when the element's style is changed.

https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle#Syntax